File: perl/pwg.pl
#!/usr/bin/env perl

use warnings;
use strict;
$|++;

srand(time());
my $dictfile = '/usr/share/dict/words';
my %words;
my @fragment_lengths = 3..5;
my $length = 12;
my $middle_length = 3;
my @specials = ('!', '@', '#', '$', '%', '^', '&', '*', '(', ')');


sub read_words {
    my @rawwords;
    my $f;
    open($f, "<$dictfile") or die "Cannot open $dictfile: $!\n";
    @rawwords = map {
        chomp;
        lc;
    } grep {
        /^[a-zA-Z]{$fragment_lengths[0],$fragment_lengths[-1]}?$/
    } <$f>;
    close($f);
    foreach my $fragment_length (@fragment_lengths) {
        @{$words{$fragment_length}} = grep { (length($_) == $fragment_length) } @rawwords;
    }
}

sub middle {
    my $length = shift || $middle_length;
    my $symbol = $specials[int(rand($#specials+1))];
    my $digits;
    foreach my $i (0..($length - 2)) {
        $digits .= int(rand(10));
    }
    return $digits . $symbol;
}

sub generate {
    my $count = shift || 1;
    my @passwords;
    foreach my $i (0..($count - 1)) {
        my $start_length = $fragment_lengths[int(rand($#fragment_lengths))];
        my $midlen = $middle_length;
        if ($start_length == 3) {
            $midlen ++;
        }
        my $end_length = $length - ($start_length + $midlen);
        my $start = @{$words{$start_length}}[int(rand($#{$words{$start_length}}))];
        $start = ucfirst($start) if (int(rand(10) % 2) == 0);
        my $middle = middle($midlen);
        my $end = @{$words{$end_length}}[int(rand($#{$words{$end_length}}))];
        $end = ucfirst($end) if (int(rand(10) % 2) == 0);
        push(@passwords, $start . $middle . $end);
    }
    return @passwords;
}


my $count = shift || 1;
read_words();
my @passwords = generate($count);
map {print $_, "\n"} @passwords;

# vim:filetype=perl

File: ruby/pwg.rb
#!/usr/bin/env ruby

class PasswordGenerator
  def initialize()
    srand Time.now.to_i
    @words = Hash.new
    @lengths = (4..5)
    @specials = ['!', '@', '#', '$', '%', '^', '&', '*', '(', ')', '[', ']', '{', '}']
    read_words
  end

  def generate()
    length = @lengths.to_a[rand(@lengths.to_a.length)]
    word1 = @words[length][rand(@words[length].length)]
    word2 = @words[length][rand(@words[length].length)]
    mid = ''
    (word1.length + word2.length).to_i.upto(10) do |count|
      mid = mid + rand(9).to_s
    end
    mid = mid + @specials[rand(@specials.length)]
    if (rand(10) % 2 == 0)
      word1.capitalize!
    end
    if (rand(10) % 2 == 0)
      word2.capitalize!
    end
    word = word1 + mid + word2
  end

  private

  def read_words()
    File.open('/usr/share/dict/words') do |file|
      file.each do |word|
        word.chomp!
        word.downcase!
        if @lengths.include? word.length
          if not @words.has_key? word.length
            @words[word.length] = Array.new
          end
          @words[word.length].push(word)
        end
      end
    end
  end

end

if ARGV.length < 1
  count = 1
else
  count = ARGV[0].to_i
end
pwg = PasswordGenerator.new
1.upto(count) do
  puts pwg.generate
end
